base_agent/documents/logical_form_specification/events.md

```diff
@@ -1,10 +1,115 @@
+
+In pseudocode:
+
+```python
+started = False
+finished = False
+while True: #agent loop:
+    # this block is executed independently for each not finished command:
+    if not started:
+        if init_condition:
+            started = True
+    if started and not finished 
+        if run_condition:
+	    agent_set_action_sequence_priority_to_runme
+	if stop_condition:
+	    agent_set_action_sequence_priority_to_check
+	    
+    if remove_condition:
+        finished = True
+```
+
+Each condition key and the "spatial_control" blocks are optional. 
+The defaults (if a key is not set) are "init_condition": ALWAYS, "run_condition": ALWAYS, "stop_condition": NEVER.
+
+The default for "remove_condition" is a bit more involved. Note that conditions are checked at each agent.step(),
+and an "ALWAYS" would then immediately remove the command.  While we can check conditions on the task, the FILTERS
+for the task sequence would be different for every command... so we introduce a special value for use inside
+a CONDITION in a "control" block: "this_sequence", referring to the command sequence that is a sibling of the "control" key.
+The FILTER corresponding to "this_sequence" is
+```
+THIS_ACTION_SEQUENCE = {"memory_type": "TASKS",
+                        "triples": [{pred_text:"has_tag", "obj_text": "THIS_SEQUENCE"}]}
+
 ```
+(we will need to update this when we have "obj_text": {"fixed_value": "THIS_SEQUENCE"}).  Furthermore, for each Task in memory, we store the
+number of times it has run; and so have the attribute "RUN_COUNT".  The default (that is, what would be interpreted
+if the key is not set) for remove_condition is 
+```
+THIS_SEQUENCE_FINISHED = {
+                           "comparator": {
+			     "comparison_type": "EQUAL",
+                             "input_left": {
+			       "value_extractor": {
+			         "filter": {
+			           "memory_type": "TASKS",
+				   "triples": [{pred_text:"has_tag", "obj_text": "THIS_SEQUENCE"}],
+				   "output": {"attribute": "RUN_COUNT"}
+				 }
+			       }
+			     },
+			     "input_right": {
+			       "value_extractor": {
+			         "fixed_value": 1
+			       }
+			     }
+			    }
+			   }
+			  }
+```
+
+For "repeat 10 times" the remove condition is
+```
+{
+  "comparator": {
+    "comparison_type": "EQUAL",
+    "input_left": {
+      "value_extractor": {
+	"filter": {
+	  "memory_type": "TASKS",
+	  "triples": [{pred_text:"has_tag", "obj_text": "THIS_SEQUENCE"}],
+	  "output": {"attribute": "RUN_COUNT"}
+	}
+      }
+    },
+    "input_right": {
+      "value_extractor": {
+	"span": [0, [1,1]]
+      }
+    }
+  }
+}
+
+```
+
+We don't _really_ need 4 conditions; and could probably merge run_condition and stop_conditions, but separating allows each of
+the conditions to be simpler (for example when the condition for running is different from the condition for stopping).
+
+
+TODO modifiers (faster/slower, happily, etc...) 
```

Comment:
this MD is useful, thanks! Reminder to generate feature-level tasks for follow-ups after this.